### PR TITLE
feature: give contracts the ability to pause invitations

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
refs: #5036

## Description

Give contracts the ability to pause invitations

This is the basic capability. The remainder of the task is making it
possible for governance to manage it.

### Security Considerations

Without more and better controls, this could make it possible for contract authors to wedge their contracts. Fortunately, the next step is adding some support in contractGovernance so this power is harder to abuse or misuse.

### Documentation Considerations

This should be added to Zoe documentation, though it's not a compatibility issue.

### Testing Considerations

Tested matching patterns, non-matching patterns, and no patterns. Tested that a matched pattern prevented exercising an offer.

### CODE REVIEW

**Comment by Chris-Hibbert** [comment]
@turadg   In local testing, `test-priceAggregator.js` (oracle invitation) failed with **Promise returned by test never resolved**. I'd appreciate any help you can offer on narrowing this down. 

I've done some logging and single stepping, but that hasn't made me feel closer to identifying the problem

File: packages/zoe/src/zoeService/startInstance.js
Comment (turadg) [inline packages/zoe/src/zoeService/startInstance.js]: ASCII has more than these symbols. 
```suggestion
const isAlphaNumeric = /^[a-zA-Z0-9_]+$/;
```

File: packages/zoe/src/zoeService/startInstance.js
Comment (turadg) [inline packages/zoe/src/zoeService/startInstance.js]:
```suggestion
        X`Snoozed strings (${q(strings)}) must be non-empty alphanumeric strings.`,
```

File: packages/zoe/src/zoeService/offer/offer.js
Comment (turadg) [inline packages/zoe/src/zoeService/offer/offer.js]: this reads like the particular offer is being rejected and another might be accepted.

shouldn't it use details/quote?
```suggestion
      X`contract is not currently accepting offers. ${q(description)}`,
```

File: packages/zoe/src/internal-types.js
Comment (turadg) [inline packages/zoe/src/internal-types.js]: "snoozed" to me evokes "pause notifications". When you snooze an alarm, the time keeps ticking but it doesn't make sounds.

This is different in that notifications still come out but no offers are _accepted_. I like metaphors but here it might be better to be perfectly explicit: `isAcceptingOffers` (the inverse). It parallels `stopAcceptingOffers`.

File: packages/zoe/src/zoeService/startInstance.js
Comment (turadg) [inline packages/zoe/src/zoeService/startInstance.js]: Looks like this is to replace the contents of the array. I'm unfamiliar with this idiom. Is it common in the repo?

What's the advantage vs `let snoozedStrings` above and `snoozedStrings = strings` here? I guess mutability. But we could type the array as `Readonly<string[]>` to give confidence it won't be modified.

File: packages/zoe/src/zoeService/startInstance.js
Comment (turadg) [inline packages/zoe/src/zoeService/startInstance.js]: not clear to me why any substring match is sufficient. please explain in jsdoc comment.

File: packages/zoe/src/zoeService/offer/offer.js
Comment (Chris-Hibbert) [reply packages/zoe/src/zoeService/offer/offer.js]: It is correct that it is accepting some and rejecting others.  If the pattern list covers all invitation patterns, it might reject all of them. This code doesn't check to see if the pattern list is empty, so it doesn't know that anything is being rejected, unless the current invitation is blocked.

File: packages/zoe/src/zoeService/startInstance.js
Comment (Chris-Hibbert) [reply packages/zoe/src/zoeService/startInstance.js]: Good point

File: packages/zoe/src/zoeService/startInstance.js
Comment (Chris-Hibbert) [reply packages/zoe/src/zoeService/startInstance.js]: done.

File: packages/zoe/src/zoeService/startInstance.js
Comment (Chris-Hibbert) [reply packages/zoe/src/zoeService/startInstance.js]: Done

File: packages/zoe/src/zoeService/startInstance.js
Comment (Chris-Hibbert) [reply packages/zoe/src/zoeService/startInstance.js]: > Looks like this is to replace the contents of the array. I'm unfamiliar with this idiom. Is it common in the repo?

I haven't seen it before; I found it (and other suggestions) on stackoverflow.

The advantage is immutability. `const` is enforced, while typescript is overridable. (and in fact, we'd override it in order to set the value.)

I'm open to other suggestions for maintaining a list of suppressed values.  I don't think we need regexps. I don't think we need an API that provides a lot of support. My expectation is that most contracts will leave it to governance, and that governance will have fairly rough tools, as it's only expected to be used in emergencies.

Comment (turadg) [reply packages/zoe/src/zoeService/startInstance.js]: > The advantage is immutability. const is enforced

Well sorta. The immutability of the binding is enforced, but it's an array and this line replaces the contents. So it's a full mutation. I'd say the `const` is more deceiving than helpful. It's a hoop to jump through to avoid rebinding the value.

I think it would be more direct to `let snoozedStrings`. The reader can tell at that point that it'll be reassigned. And it has the added benefit of making it possible to express that the array itself should not be mutated, with `Readonly<>`. Yes it can be overridden but at least the intent is expressed. And there's no loss to material runtime mutability.

File: packages/zoe/src/zoeService/offer/offer.js
Comment (turadg) [reply packages/zoe/src/zoeService/offer/offer.js]: ah, I misread `updateSnoozedList`. That method name could be improved.

For one, isn't it setting the list? There's no history or partial update; it's a full replacement.

For two, I maintain that "snooze" is too metaphorical for the concept of "stop accepting offers whose descriptions contain one of these strings". (Incidentally, that could make a good JSdoc for the method, if I got it right)

Three, "list" is vague. List of what? Not snoozes.  Consider "selectors". Or since it's about what to reject, "filter"

Putting it all together, maybe `setOfferFilter`. Where the filter is a list of substrings that can match.

File: packages/zoe/src/zoeService/startInstance.js
Comment (Chris-Hibbert) [reply packages/zoe/src/zoeService/startInstance.js]: done.

File: packages/zoe/src/internal-types.js
Comment (Chris-Hibbert) [reply packages/zoe/src/internal-types.js]: good suggestion.

File: packages/zoe/src/zoeService/startInstance.js
Comment (turadg) [inline packages/zoe/src/zoeService/startInstance.js]: suffices to say "must be an Array of alpha-numeric strings"

File: packages/zoe/src/zoeService/offer/offer.js
Comment (Chris-Hibbert) [reply packages/zoe/src/zoeService/offer/offer.js]: done

File: packages/zoe/src/zoeService/startInstance.js
Comment (Chris-Hibbert) [reply packages/zoe/src/zoeService/startInstance.js]: I want to point out that that's how you reset the list.

Comment (turadg) [reply packages/zoe/src/zoeService/startInstance.js]: pointing out how to reset the list is good. I'd expect to see that in the function jsdoc.  I'd probably want to know that before I send an invalid argument :)

File: packages/zoe/src/internal-types.js
Comment (turadg) [inline packages/zoe/src/internal-types.js:83]: 👍

File: packages/zoe/src/zoeService/offer/offer.js
Comment (turadg) [inline packages/zoe/src/zoeService/offer/offer.js]: does this need `details`?

**Review by turadg — APPROVED** [review]

File: packages/zoe/src/zoeService/offer/offer.js
Comment (Chris-Hibbert) [reply packages/zoe/src/zoeService/offer/offer.js]: Yes, thanks.  I added a clause to my pre-submit  pattern, and found a second case.  Doh!

**Comment by Chris-Hibbert** [comment]
@erights   I'm waiting for a review from you.

File: packages/zoe/src/zoeService/offer/offer.js
Comment (erights) [inline packages/zoe/src/zoeService/offer/offer.js]: In the case where the offer is not blocked, I would not have expected this PR to cause an observable difference. However, the old logic first burns the offer before asserting that the contract is still accepting offers. The new one asserts that the contract is still accepting offers before it burns. Arguably the new behavior is better, in which case no change suggested. But I would like to understand the implications of making this observable change before we proceed.

File: packages/zoe/src/zoeService/startInstance.js
Comment (erights) [inline packages/zoe/src/zoeService/startInstance.js]: The original is ok. I'll explain why I suggest each of the changes below, but take only what you agree is an improvement.

```suggestion
    let offerFilterStrings = harden([]);
    // if any string in the list matches, don't process the invitation
    const isBlocked = string =>
      offerFilterStrings.some(s => string.indexOf(s) > -1);

    const setOfferFilter = strings => {
      harden(strings);
      fit(strings, M.arrayOf(M.string()));
      assert(
        strings.every(s => isAlphanumeric.test(s)),
        X`Blocked strings (${q(
          strings,
        )}) must be a (possibly empty) Array of alpha-numeric strings.`,
      );

      offerFilterStrings = strings;
```

I see a declaration of `ReadOnly` and code consistent with the array actually being immutable. The original copies the incoming array into a new array which it then never mutates. So it seems stronger than ReadOnly anyway.

But TS is unsound. If you want to ensure that the array isn't mutated, you can harden it. Fortunately, a hardened array of strings is a CopyArray, and so a Passable, and so something we can pattern match on. The pattern match above asserts in one line what took several lines to say. Unfortunately, we cannot express the alphanumeric constraint in the pattern language so that still needs to be explicit.

A possible reason to prefer your code over my suggested pattern match: Better diagnostics. My pattern matching logic doesn't understand the purpose of the match, and so cannot give diagnostics that are as informative.

Once we know that `strings` is a CopyArray of strings, we know it is safe enough that we don't need to copy it into `proposedStrings`. OTOH, this hardens whatever `strings` argument was passed in by our caller, which we might want to avoid. In that case, we can start with
```js
const proposedStrings = harden([...strings]);
```
and then do everything else with `proposedStrings` more like your original. Note that `[...foo]` is essentially the same as `Array.from(foo)` but IMO reads nicer.

File: packages/zoe/src/zoeService/startInstance.js
Comment (erights) [inline packages/zoe/src/zoeService/startInstance.js]: This is the part I don't get: Why substring rather than exact match? Do you have a use case in mind?

File: packages/zoe/src/zoeService/startInstance.js
Comment (erights) [reply packages/zoe/src/zoeService/startInstance.js]: I did not change the code for `isBlocked` above because I discuss the substring match criteria separately.

File: packages/zoe/test/unitTests/contracts/test-priceAggregator.js
Comment (erights) [inline packages/zoe/test/unitTests/contracts/test-priceAggregator.js:11]: What are the changes to this file about? They seem completely unrelated to the rest of this PR.

File: packages/zoe/test/unitTests/test-blockedOffers.js
Comment (erights) [inline packages/zoe/test/unitTests/test-blockedOffers.js]: That setting the offer filter to `['eat']` blocks invitations with description `'seat'` is disturbing. It is clearly accident prone. For the scenarios off the top of my head, exact match is fine. Why do you want substring? Do you have a use case in mind?

(I seem to have a vague memory that we may have discussed this and I agreed with substring. But I don't quite remember and I am confused now.)

**Review by erights — CHANGES_REQUESTED** [review]
I see that you make `setOfferFilter` available on `zcf` and thereby callable by the contract without requiring any governance decision. This raises the invitation legibility question. Tell me if the following is the answer:

In order for the contract to call `zcf.setOfferFilter(['exerciseOption'])` it would either need to say this statically where Fred could see it, or engage in some obscure code that Fred would never normally see in a contract that prevents Fred from statically determining that this isn't called?

This does not technically violate invitation legibility, but does add to Fred's work in a way that's hard to explain. Are you sure you want to do this?

File: packages/zoe/src/zoeService/offer/offer.js
Comment (Chris-Hibbert) [reply packages/zoe/src/zoeService/offer/offer.js]: I misunderstood the semantics of `assertAcceptingOffers()`. Since Zoe does not stop accepting offers and then restart, it's legitimate to burn the offer before checking. If we're no longer accepting offers, then the invitation won't be useable later.

File: packages/zoe/src/zoeService/startInstance.js
Comment (Chris-Hibbert) [reply packages/zoe/src/zoeService/startInstance.js]: My mental model was a big complex contract. If the author can design the contract so some subset of invitations can be disabled as a group with a single string, that makes it easier to manage.

My theory was that regExps are expensive, and we wouldn't want contracts specifying patterns for Zoe to execute. 

My usual advice would be to specify full strings that match particular invitations, and to be very careful if they're thinking about using short strings. With most contracts, the author should be able to quickly look at all the invitation descriptions. 

I'm open to other approaches, but if a contract has 10 invitation descriptions (or more), it wouldn't be a great API to require them to list all of them in order to stop the contract.

File: packages/zoe/src/zoeService/startInstance.js
Comment (Chris-Hibbert) [reply packages/zoe/src/zoeService/startInstance.js]: Rather than side-effect the caller-provided string, I used your suggestion for `harden([...strings])`. 

I do prefer the better error message, so I bundled the tests together into `isAlphaNumeric()`;

File: packages/zoe/test/unitTests/contracts/test-priceAggregator.js
Comment (Chris-Hibbert) [reply packages/zoe/test/unitTests/contracts/test-priceAggregator.js:11]: I think this is a consequence of adding ` await query.getInvitationDetails()` to the top of offer(). I don't see an alternative to doing that.

If pausing weren't such a crucial feature for protecting against runaway contracts in an emergency, I would be sympathetic to an argument that the required extra `await` is too high a price to pay for this feature.

File: packages/zoe/test/unitTests/test-blockedOffers.js
Comment (Chris-Hibbert) [reply packages/zoe/test/unitTests/test-blockedOffers.js]: short strings in the block list are a bad idea, unless the goal is to block everything.  I talked about my reasoning for substrings in `startInstance.js`.

I'll change the test to not use a short string, since I agree that it's a bad example.

File: packages/zoe/test/unitTests/contracts/test-priceAggregator.js
Comment (erights) [reply packages/zoe/test/unitTests/contracts/test-priceAggregator.js:11]: We discussed. It is just extra intra-vat turns. No extra cranks, blocks, or network round trips. Should be assumed to have negligible performance implications until there is evidence otherwise. Resolving.

File: packages/zoe/src/zoeService/startInstance.js
Comment (erights) [inline packages/zoe/src/zoeService/startInstance.js]:
```suggestion
    // Invitations whose descriptions match any of the strings will be blocked
```

File: packages/zoe/src/zoeService/startInstance.js
Comment (erights) [reply packages/zoe/src/zoeService/startInstance.js]: Chris and I discussed. No special constraints on these strings are needed, and in any case would need to be coordinated with similar constraints on the descriptions strings, which are equally unneeded.

Gone. Resolving.

File: packages/zoe/src/zoeService/startInstance.js
Comment (erights) [reply packages/zoe/src/zoeService/startInstance.js]: substring match is gone. Resolving.

File: packages/zoe/src/zoeService/startInstance.js
Comment (erights) [reply packages/zoe/src/zoeService/startInstance.js]: Restrictions on the string contents are gone. Resolving.

File: packages/zoe/src/zoeService/offer/offer.js
Comment (erights) [reply packages/zoe/src/zoeService/offer/offer.js]: I am sooo looking forward to a lint checking rule that catches these missing `X`s

File: packages/zoe/src/zoeService/offer/offer.js
Comment (erights) [reply packages/zoe/src/zoeService/offer/offer.js]: Original order restored. Resolving.

File: packages/zoe/src/zoeService/startInstance.js
Comment (erights) [reply packages/zoe/src/zoeService/startInstance.js]: Current state looks good. Resolving.

**Review by erights — APPROVED** [review]
LGTM!

File: packages/zoe/src/zoeService/startInstance.js
Comment (Chris-Hibbert) [reply packages/zoe/src/zoeService/startInstance.js]: yep. that's better.

**Pull request merged**
